packages/core/src/provider.tsx

```diff
@@ -1,27 +1,27 @@
 import type { Networkish } from '@ethersproject/networks'
 import type { BaseProvider, Web3Provider } from '@ethersproject/providers'
 import type { Connector } from '@web3-react/types'
-import type { ReactNode } from 'react'
+import type { Context, ReactNode } from 'react'
 import React, { createContext, useContext } from 'react'
 import type { Web3ReactHooks, Web3ReactPriorityHooks } from './hooks'
 import { getPriorityConnector } from './hooks'
 
-type Web3ContextType = {
+export type Web3ContextType<T extends BaseProvider = Web3Provider> = {
   connector: ReturnType<Web3ReactPriorityHooks['usePriorityConnector']>
   chainId: ReturnType<Web3ReactPriorityHooks['usePriorityChainId']>
   accounts: ReturnType<Web3ReactPriorityHooks['usePriorityAccounts']>
   isActivating: ReturnType<Web3ReactPriorityHooks['usePriorityIsActivating']>
   error: ReturnType<Web3ReactPriorityHooks['usePriorityError']>
   account: ReturnType<Web3ReactPriorityHooks['usePriorityAccount']>
   isActive: ReturnType<Web3ReactPriorityHooks['usePriorityIsActive']>
-  provider: ReturnType<Web3ReactPriorityHooks['usePriorityProvider']>
+  provider: Web3Provider | undefined | T
```

Comment:
I think this needs to be `T | undefined`, with the `Web3Provider` brought in by `T`.

Reply:
I think that `useSelectedProvider` (the eventual root of this type definition) should also be more explicit. Right now it returns `Web3Provider | T | undefined`, but it should return `T | undefined` by casting the return value: `return values[index] as T | undefined`.